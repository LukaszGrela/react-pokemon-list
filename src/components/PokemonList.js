import React from 'react';
import PropTypes from 'prop-types';
import { API_GET_SPRITE_FRONT } from '../constants/api';

import './styles/PokemonList.scss';

export default class PokemonList extends React.Component {
    render = () => {
        const { list, handleClick } = this.props;
        return (
            <ul className={'pokemon-list' + (handleClick ? ' interactive' : '')}>
                {
                    list.map(({ url, name, id }, index) => <li key={index}
                        onClick={() => {
                            handleClick && handleClick(id);
                        }}
                        className='pokemon-item'>
                        <img src={API_GET_SPRITE_FRONT(id)} className='pokemon-image'
                            onError={(e) => {
                                e.target.src = API_GET_SPRITE_FRONT('default/0');
                            }} />
                        <div className='pokemon-name'>{name}</div>
                    </li>)
                }
            </ul>)

    }
}

PokemonList.propTypes = {
    list: PropTypes.arrayOf(PropTypes.object).isRequired,
    handleClick: PropTypes.func
}