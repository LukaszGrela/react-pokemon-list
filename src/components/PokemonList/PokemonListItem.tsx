import React from 'react';

export interface IProps {
  n: number;
  pid: string;
  name: string;
}

const PokemonListItem: React.FC<IProps> = ({
  n,
  pid,
  name,
}: IProps): JSX.Element => {
  return (
    <li
      className='PokemonListItem'
      onClick={() => {
        window.location.hash = `/item/${pid}`;
        console.log(window.location.hash);
      }}
    >
      <span className='number'>#{n}</span>

      {/* <Image
    src={API_GET_SPRITE_FRONT(pid)}
    fallback={API_GET_SPRITE_FRONT('default/0')}
    ref="Image"
    alt={`Image of ${name} pokemon.`} /> */}

      <div className='PokemonListItem_name'>{name}</div>
    </li>
  );
};

export default PokemonListItem;
