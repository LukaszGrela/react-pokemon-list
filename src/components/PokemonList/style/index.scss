@import 'styles/exports';

@import './PokemonList';
