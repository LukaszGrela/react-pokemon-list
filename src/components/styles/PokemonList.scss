
.pokemon-list {

    .pokemon-item {
        list-style: none;
        width: 10.4rem;
        background-color: #f7f7f7;
        text-align: center;
        margin: 1rem;
        border-radius: 0.8rem;
        display: inline-block;

        img {
            width: 9.6rem;
            height: 9.6rem;
            margin: 0.4rem;
        }

        .pokemon-name {
            text-transform: capitalize;
            font-weight: 700;
            font-family: sans-serif;
            margin: 0.4rem 0;
            width: 10.4rem;
        }

    }
}