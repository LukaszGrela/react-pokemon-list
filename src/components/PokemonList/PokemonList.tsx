import React, { ReactNode } from 'react';
import { PAGINATION, parseIdFromUrl } from '../../api';
import type { TNamedAPIResource } from '../../store/model/common';
import { useGetPokemonsListQuery } from '../../store/services/pokemons-list';
import PokemonListItem from '../PokemonListItem/PokemonListItem';

import './style/index.scss';
export interface IProps {
  page: number;

  interactive: boolean;

  handlePokemonSelect: (pid: string, name: string) => void;
}

const PokemonList: React.FC<IProps> = (props: IProps): JSX.Element => {
  const { page, interactive, handlePokemonSelect } = props;
  const { data: results, error, isLoading: loading, } = useGetPokemonsListQuery(PAGINATION(page));

  return (
    <>
      {loading && <div className='PokemonList PokemonList_loading'>Loading...</div>}
      {!loading && (
        <ul className={`PokemonList${interactive ? ' interactive' : ''}`}>
          {results?.results.map(
            (resource: TNamedAPIResource): ReactNode => {
              console.log(resource)
              const key = resource.url;
              return (
                <PokemonListItem
                  key={key}
                  pid={`${parseIdFromUrl(resource.url) || ''}`}
                  name={resource.name}
                  onClick={(pid) => handlePokemonSelect(pid, resource.name)}
                />
              );
            }
          )}
        </ul>
      )}
    </>
  );
};

export default PokemonList;
