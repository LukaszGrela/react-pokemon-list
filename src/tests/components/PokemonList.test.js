import React from 'react';
import {shallow} from 'enzyme';
import PokemonList from '../../components/PokemonList';
// import {POKEMON_3 } from '../fixtures/pokemon-list.js';


test('Should render component correctly', () => {
    const wrapper = shallow(<PokemonList 
        list={
            [
                { "url": "https:\/\/pokeapi.co\/api\/v2\/pokemon\/21\/", "name": "spearow", id:21 }
            ]
        }
        />);

    expect(wrapper).toMatchSnapshot();
})