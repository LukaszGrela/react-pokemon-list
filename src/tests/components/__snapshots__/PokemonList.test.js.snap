// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render component correctly 1`] = `
<ul
  className="pokemon-list"
>
  <li
    className="pokemon-item"
    key="0"
    onClick={[Function]}
  >
    <img
      className="pokemon-image"
      onError={[Function]}
      src="./static/pokemon/21.png"
    />
    <div
      className="pokemon-name"
    >
      spearow
    </div>
  </li>
</ul>
`;
